### ISSUE TITLE
Regardless of the number of gpus, the memory usage of all gpus is the same. and memory usage is always high

### ISSUE THREAD

**Issue by Ladun** [microsoft/DeepSpeed#1883]
**Describe the bug**
I am going to try the ZeRO optimizer with huggingface Trainer on stage 2 and stage 3.
- First, some weird things is regardless of the number of gpus, the memory usage of all gpus is the same. This phenomenon is the same for stage 2 or 3.
- Second, strange things is that stage3 is much slower than stage2.
- Last, gpu memory usage is always high.

I use the `EleutherAI/gpt-j-6b` model.

This is my training code
```
...
train_args = TrainingArguments(
       ...,
       deepspeed=deepspeed_config_file,
       fp16=True
)
...
tokenizer = AutoTokenizer.from_pretrained(model_name_or_path, use_fast_tokenizer=False)
model     = GPTJForSequenceClassification.from_pretrained(model_name_or_path)
...
trainer = Trainer(
        model=model,
        tokenizer=tokenizer,
        ...
        args=train_args
    )
...
```

<details> 
<summary>Config for stage 2</summary> 

```{
    "fp16": {
        "enabled": "auto",
        "loss_scale": 0,
        "loss_scale_window": 1000,
        "initial_scale_power": 12,
        "hysteresis": 2,
        "min_loss_scale": 1
    },
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": "auto",
            "betas": "auto",
            "eps": "auto",
            "weight_decay": "auto"
        }
    },
    "scheduler": {
        "type": "WarmupLR",
        "params": {
            "warmup_min_lr": "auto",
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto"
        }
    },
    "zero_optimization": {
	     "stage": 2,
	     "allgather_partitions": true,
	     "allgather_bucket_size": 2e8,
	     "reduce_scatter": true,
	     "reduce_bucket_size": 2e8,
	     "overlap_comm": true,
	     "contiguous_gradients": true,
	     "cpu_offload": true
	  },
    "gradient_accumulation_steps": "auto",
    "gradient_clipping": "auto",
    "steps_per_print": 2000,
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto",
    "wall_clock_breakdown": false
}
```

 </details>

<details> 
<summary>Config for stage 3</summary> 

```
{
    "fp16": {
        "enabled": "auto",
        "loss_scale": 0,
        "loss_scale_window": 1000,
        "initial_scale_power": 12,
        "hysteresis": 2,
        "min_loss_scale": 1
    },
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": "auto",
            "betas": "auto",
            "eps": "auto",
            "weight_decay": "auto"
        }
    },
    "scheduler": {
        "type": "WarmupLR",
        "params": {
            "warmup_min_lr": "auto",
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto"
        }
    },
    "zero_optimization": {
        "stage": 3,
        "offload_optimizer": {
            "device": "cpu",
            "pin_memory": true
        },
        "offload_param": {
            "device": "cpu",
            "pin_memory": true
        },
        "overlap_comm": true,
        "contiguous_gradients": true,
        "sub_group_size": 1e8,
        "reduce_bucket_size": "auto",
        "stage3_prefetch_bucket_size": "auto",
        "stage3_param_persistence_threshold": "auto",
        "stage3_max_live_parameters": 1e9,
        "stage3_max_reuse_distance": 1e9,
        "stage3_gather_fp16_weights_on_model_save": true
    },
    "gradient_accumulation_steps": "auto",
    "gradient_clipping": "auto",
    "steps_per_print": 2000,
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto",
    "wall_clock_breakdown": false
}
```

 </details>

When i use the 4 gpus, memory usage in stage 2 is as follows:

GPU0 (A100) -> [39328M / 40536M]
GPU1 (A100) -> [39512M / 40536M]
GPU2 (A100) -> [38973M / 40536M]
GPU3 (A100) -> [39123M / 40536M]

When i use the 1 gpus, memory usage in stage 2 is as follows:
GPU0 (A100) -> [39328M / 40536M]

Memory usage in Stage 3 is reduced, but as above, regardless of the number of gpus, it is the same every time, and the learning speed is significantly reduced.
**Expected behavior**

If the number of gpu is increased, I thought that memory usage would decrease. but it doesn't

**ds_report output**
Please run `ds_report` to give us details about your setup.
<img width="1063" alt="스크린샷 2022-04-06 오후 2 57 26" src="https://user-images.githubusercontent.com/47883234/161905480-f5c93e31-1e37-478f-8383-8cafd52e7773.png">

**Screenshots**

**System info (please complete the following information):**
 - OS: [e.g. Ubuntu 20.04.4 LTS]
 - GPU count and types [e.g. one machines with x4 A100s]
 - Python 3.9.7
 - Any other relevant info about your setup

**Launcher context**
Are you launching your experiment with the `deepspeed` launcher, MPI, or something else?

I am launching with the`deepspeed`

**Comment by mrwyattii** [comment]
I believe this is expected behavior given the size/memory requirements of the `gpt-j-6b` model:
```bash
❯ python -c 'from transformers import AutoModel; \
from deepspeed.runtime.zero.stage_1_and_2 import estimate_zero2_model_states_mem_needs_all_live; \
model = AutoModel.from_pretrained("EleutherAI/gpt-j-6B"); \
estimate_zero2_model_states_mem_needs_all_live(model, num_gpus_per_node=1, num_nodes=1)
...
HW: Setup with 1 node, 1 GPU per node.
SW: Model with 5844M total params.
  per CPU  |  per GPU |   Options
  130.63GB |  10.89GB | offload_optimizer=cpu
   32.66GB | 108.86GB | offload_optimizer=none
❯ python -c 'from transformers import AutoModel; \
from deepspeed.runtime.zero.stage_1_and_2 import estimate_zero2_model_states_mem_needs_all_live; \
model = AutoModel.from_pretrained("EleutherAI/gpt-j-6B"); \
estimate_zero2_model_states_mem_needs_all_live(model, num_gpus_per_node=4, num_nodes=1)
...
HW: Setup with 1 node, 4 GPUs per node.
SW: Model with 5844M total params.
  per CPU  |  per GPU |   Options
  130.63GB |  10.89GB | offload_optimizer=cpu
  130.63GB |  43.54GB | offload_optimizer=none
```

**Comment by Ladun** [comment]
Test setting is 
- use "EleutherAI/gpt-j-6B"
- batch size 4
- use gpu A100


When I test it again, it looks like this:
    On 1 GPU
<img width="574" alt="스크린샷 2022-04-07 오전 10 38 01" src="https://user-images.githubusercontent.com/47883234/162103831-7cb1181f-47a9-4832-a778-787065fa829b.png">

On 4 GPUs
<img width="556" alt="스크린샷 2022-04-07 오전 10 45 55" src="https://user-images.githubusercontent.com/47883234/162103845-32a7b3c4-c77b-4fc1-9534-62eb225f62fd.png">

I also checked by running "estimate_zero2_model_states_mem_needs_all_live", but in reality I get different results, so I write a question.

**Comment by Ladun** [comment]
> I believe this is expected behavior given the size/memory requirements of the `gpt-j-6b` model:
> 
> ```shell
> ❯ python -c 'from transformers import AutoModel; \
> from deepspeed.runtime.zero.stage_1_and_2 import estimate_zero2_model_states_mem_needs_all_live; \
> model = AutoModel.from_pretrained("EleutherAI/gpt-j-6B"); \
> estimate_zero2_model_states_mem_needs_all_live(model, num_gpus_per_node=1, num_nodes=1)
> ...
> HW: Setup with 1 node, 1 GPU per node.
> SW: Model with 5844M total params.
>   per CPU  |  per GPU |   Options
>   130.63GB |  10.89GB | offload_optimizer=cpu
>    32.66GB | 108.86GB | offload_optimizer=none
> ❯ python -c 'from transformers import AutoModel; \
> from deepspeed.runtime.zero.stage_1_and_2 import estimate_zero2_model_states_mem_needs_all_live; \
> model = AutoModel.from_pretrained("EleutherAI/gpt-j-6B"); \
> estimate_zero2_model_states_mem_needs_all_live(model, num_gpus_per_node=4, num_nodes=1)
> ...
> HW: Setup with 1 node, 4 GPUs per node.
> SW: Model with 5844M total params.
>   per CPU  |  per GPU |   Options
>   130.63GB |  10.89GB | offload_optimizer=cpu
>   130.63GB |  43.54GB | offload_optimizer=none
> ```
> 
> What is your system memory usage when running on 1 vs 4 GPUs?

Test setting is 
- use "EleutherAI/gpt-j-6B"
- batch size 4
- use gpu A100


When I test it again, it looks like this:
    On 1 GPU
<img width="574" alt="스크린샷 2022-04-07 오전 10 38 01" src="https://user-images.githubusercontent.com/47883234/162103831-7cb1181f-47a9-4832-a778-787065fa829b.png">

On 4 GPUs
<img width="556" alt="스크린샷 2022-04-07 오전 10 45 55" src="https://user-images.githubusercontent.com/47883234/162103845-32a7b3c4-c77b-4fc1-9534-62eb225f62fd.png">

I also checked by running "estimate_zero2_model_states_mem_needs_all_live", but in reality I get different results, so I write a question.

**Comment by Ladun** [comment]
> I think the batch size you are setting is per-GPU, which would explain why GPU memory usage does not decrease as you increase the number of GPUs. The other thing affecting the GPU memory usage you see is Zero-offloading. With Stage 2, optimizer states and gradients are partitioned across processes. Using this with Zero-offloading will offload to CPU memory. See the table below, which shows GPU memory for 1 and 2 GPUs, stage 2, with and without offloading enabled (these numbers are from a different transformer model, on 2xA6000):
> 
> # GPUs	Zero-offload	Global BS	GPU Memory Usage
> 1	no	16	(44GB, 00GB)
> 2	no	16	(28GB, 29GB)
> 1	yes	16	(34GB, 00GB)
> 2	yes	16	(17GB, 17GB)
> 2	yes	32	(34GB, 40GB)
> Also note that if we increase the global batch size to 32 (or 16/gpu) as I believe you are doing, the memory usage stays the same compared to single GPU with global batch size 16 (or 16/gpu).

Thank you for your response. I'm sorry for the late reply.

I agree with you that the same memory usage due to batch_size.
